### PR TITLE
Stream performance improvements

### PR DESCRIPTION
**Pull request by tom95858**
These changes significantly reduce the overhead of publishing
JSON data if that data is being forwarded and not otherwise consumed
at the subscriber.

The publisher is also acknowledged before delivering the data to the
subscribers. This reduces the latency at the producer waiting
for the ACK.

### CODE REVIEW

**Comment by nichamon** [comment]
@tom95858 Could you please rebase this on top of the latest OVIS-4? There are many changes I believe are unrelated to the patch.

**Comment by tom95858** [comment]
Sure, testing now...

File: ldms/src/ldmsd/ldmsd_request.c [hunk 1]
```diff
@@ -6012,50 +6009,38 @@ static int stream_publish_handler(ldmsd_req_ctxt_t reqc)
 		goto err_reply;
 	}
 
+	reqc->errcode = 0;
+	ldmsd_send_req_response(reqc, "ACK");
```
Comment (nichamon) [inline ldms/src/ldmsd/ldmsd_request.c:6013]: LDMSD will send two responses back to the requester if no data is provided. Should we validate the stream before sending the ACK response (we won't validate if the JSON-formated string can be parsed or not)?

File: ldms/src/ldmsd/ldmsd_stream.c [hunk 2]
```diff
@@ -44,57 +46,90 @@ struct ldmsd_stream_s {
...
 			goto err_2;
-		}
 		pthread_mutex_init(&s->s_lock, NULL);
 		LIST_INIT(&s->s_c_list);
 		rbn_init(&s->s_ent, (char *)s->s_name);
+		pthread_mutex_lock(&s_tree_lock);
 		rbt_ins(&s_tree, &s->s_ent);
-	} else {
-		s = container_of(rbn, struct ldmsd_stream_s, s_ent);
+		pthread_mutex_unlock(&s_tree_lock);
```
Comment (nichamon) [inline ldms/src/ldmsd/ldmsd_stream.c:131]: `pthread_mutex_lock(&s->s_lock)` is missing after the call to `pthread_mutex_unlock(&s_tree_lock)`.

**Review by nichamon — CHANGES_REQUESTED** [review]

File: ldms/src/ldmsd/ldmsd_stream.c (hunk 2 above)
Comment (tom95858) [reply ldms/src/ldmsd/ldmsd_stream.c:131]: This is the path where the stream is newly allocated and therefore not locked. Am I missing something?

File: ldms/src/ldmsd/ldmsd_request.c (hunk 1 above)
Comment (tom95858) [reply ldms/src/ldmsd/ldmsd_request.c:6013]: The purpose of the change is to reduce the overhead of streams data if it is only being forwarded and not delivered. If we validate the data, ldmsd will incur the CPU hit even it is not delivering the data to any client. Good catch on the dup response. I'll fix it.

File: ldms/src/ldmsd/ldmsd_stream.c (hunk 2 above)
Comment (tom95858) [reply ldms/src/ldmsd/ldmsd_stream.c:131]: Ah...ok, I misread your comment.

**Pull request merged**